### PR TITLE
Fix "many" in fold and parser

### PR DESCRIPTION
**Pull request by harendra-kumar**
Initialize the inner fold/parser only after the first input. Like we do in foldMany.

### CODE REVIEW

File: core/src/Streamly/Internal/Data/Fold/Type.hs [hunk 1]
```diff
@@ -1628,46 +1628,39 @@ many (Fold sstep sinitial sextract) (Fold cstep cinitial cextract) =
...
+            Partial ss1 -> return $ Partial $ ManyLoop ss1 cs
+            Done sb -> doInit (cstep cs sb)
 
     {-# INLINE step #-}
-    step (ManyFirst ss cs) a = step_ ss cs a
+    step (ManyInit cs) a = do
+        sres <- sinitial
+        case sres of
+            Partial ss -> step_ ss cs a
+            Done _ -> error "many: first fold must not terminate without input"
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Fold/Type.hs:1652]: Why shouldn't it?

File: core/src/Streamly/Internal/Data/Fold/Type.hs [hunk 2]
```diff
@@ -1628,46 +1628,39 @@ many (Fold sstep sinitial sextract) (Fold cstep cinitial cextract) =
...
+    extract (ManyInit cs) = cextract cs
     extract (ManyLoop ss cs) = do
         cres <- sextract ss >>= cstep cs
         case cres of
             Partial s -> cextract s
             Done b -> return b
 
--- | Like many, but inner fold emits an output at the end even if no input is
--- received.
+-- | Like many, but the "first" fold emits an output at the end even if no
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Fold/Type.hs:1662]: Technically the second fold emits the output.
We can say "The initial value is emitted"

Ah, the first fold emits the output to the second fold.

File: core/src/Streamly/Internal/Data/Parser/ParserD/Type.hs [hunk 3]
```diff
@@ -922,56 +922,64 @@ alt (Parser stepL initialL extractL) (Parser stepR initialR extractR) =
...
-    extract (Tuple3' _ 0 fs) = fmap (Done 0) (fextract fs)
-    extract (Tuple3' s cnt fs) = do
-        r <- extract1 s
+    {-# INLINE step #-}
+    step (ManyInit fs) a = do
+        pres <- initial1
+        case pres of
+            IPartial ps -> step_ 1 ps fs a
+            IDone _ ->
+                error "many: parser must not terminate without input"
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserD/Type.hs:976]: "first parser"

File: core/src/Streamly/Internal/Data/Parser/ParserD/Type.hs [hunk 4]
```diff
@@ -983,7 +991,9 @@ splitMany (Parser step1 initial1 extract1) (Fold fstep finitial fextract) =
             Partial _ _ -> error "splitMany: Partial in extract"
             Continue n s1 -> do
                 assertM(n == cnt)
-                return (Continue n (Tuple3' s1 0 fs))
+                return (Continue n (ManyLoop 0 s1 fs))
+
+-- XXX The parser is always extracted even when there is no input.
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserD/Type.hs:996]: Make the changes to `splitManyPost` also? Similar to `manyPost`?

**Review by adithyaov — APPROVED** [review]

**Review by adithyaov — CHANGES_REQUESTED** [review]
Some comments.
Approved by mistake.

**Comment by harendra-kumar** [comment]
Closing this. Changed my mind about these changes.

**Pull request closed without merging**